leader-election-impl/src/main/java/com/palantir/paxos/SplittingPaxosStateLog.java

```diff
@@ -0,0 +1,122 @@
+    private final PaxosStateLog<V> legacyLog;
+    private final PaxosStateLog<V> currentLog;
+    private final Runnable markLegacyWrite;
+    private final Runnable markLegacyRead;
+    private final long cutoffInclusive;
+    private final AtomicLong leastLogEntry;
+
+    private SplittingPaxosStateLog(PaxosStateLog<V> legacyLog,
+            PaxosStateLog<V> currentLog,
+            Runnable markLegacyWrite,
+            Runnable markLegacyRead,
+            long cutoffInclusive,
+            AtomicLong leastLogEntry) {
+        this.legacyLog = legacyLog;
+        this.currentLog = currentLog;
+        this.markLegacyWrite = markLegacyWrite;
+        this.markLegacyRead = markLegacyRead;
+        this.cutoffInclusive = cutoffInclusive;
+        this.leastLogEntry = leastLogEntry;
+    }
+
+    public static <V extends Persistable & Versionable> PaxosStateLog<V> create(SplittingParameters<V> parameters) {
+        return new SplittingPaxosStateLog<>(
+                parameters.legacyLog(),
+                parameters.currentLog(),
+                parameters.markLegacyWrite(),
+                parameters.markLegacyRead(),
+                parameters.cutoffInclusive(),
+                new AtomicLong(Math.min(parameters.legacyLog().getLeastLogEntry(), parameters.cutoffInclusive())));
+    }
+
+    @Override
+    public void writeRound(long seq, V round) {
+        if (seq >= cutoffInclusive) {
+            currentLog.writeRound(seq, round);
+        } else {
+            markLegacyWrite.run();
+            legacyLog.writeRound(seq, round);
+            leastLogEntry.accumulateAndGet(seq, Math::min);
+        }
+    }
+
+    @Override
+    public byte[] readRound(long seq) throws IOException {
+        if (seq >= cutoffInclusive) {
+            return currentLog.readRound(seq);
+        } else {
+            markLegacyRead.run();
+            return legacyLog.readRound(seq);
+        }
+    }
+
+    @Override
+    public long getLeastLogEntry() {
+        return leastLogEntry.get();
+    }
+
+    @Override
+    public long getGreatestLogEntry() {
+        return currentLog.getGreatestLogEntry();
+    }
+
+    /**
+     * This implementation is a noop to ensure correctness of {@link #getLeastLogEntry()}.
+     */
+    @Override
+    public void truncate(long toDeleteInclusive) {
+        // noop
+    }
+
+    @Value.Immutable
+    abstract static class SplittingParameters<V extends Persistable & Versionable> {
+        abstract PaxosStateLog<V> legacyLog();
+        abstract PaxosStateLog<V> currentLog();
+        abstract Runnable markLegacyWrite();
+        abstract Runnable markLegacyRead();
+        abstract long cutoffInclusive();
+
+        @Value.Check
+        void cutoffEntryMustBeInCurrentLogOrEqualToNoEntry() {
+            try {
+                Preconditions.checkState(cutoffInclusive() == PaxosAcceptor.NO_LOG_ENTRY
+                                || currentLog().readRound(cutoffInclusive()) != null,
```

Comment:
This does incur a read whenever creating this object.

This is constructed once per (client, use-case). Let's go ahead with this for now I think

Reply:
discussing offline:
- Actually, we want to check that at least one entry exists in the new log after the cutoff - and this can be checked by verifying that the greatest log entry in `currentLog()` is at least the cutoff